### PR TITLE
refactor(TemplateParser): clearer error message for on* binding

### LINKED ISSUE #11756: `on*` event convention is overly greedy
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior** 
When defining an attribute of a component template, the [template parser](https://github.com/angular/angular/blob/2300c23332fe8a61427562770fb4e1911aad6e2e/modules/%40angular/compiler/src/template_parser/template_parser.ts#L905-L912) checks if a field name `startsWith` 'on', however this convention is overly greedy due to attribute fields potentially starting with 'on', yet not being event handlers. For example, `onlyCurrentMonth` or `once`, `onionsHaveLayers`, etc.

**Expected behavior**
Checker should interpret the camelcasing to only throw when field name starts with 'on' _followed by_ capital letters. (suggested regex `/^on[^a-z]/`)

* **Angular version:** 2.0.0

* **Browser:** [all ] 

* **Language:** [TypeScript 2.0.2]

### PR DESCRIPTION
**Pull request by vicb**
fixes #11756

security related

/cc @mhevery @rjamet @mprobst

### CODE REVIEW

**Review by mhevery — APPROVED** [review]

File: modules/@angular/core/test/linker/security_integration_spec.ts [hunk 1]
```diff
@@ -24,12 +23,21 @@ class SecuredComponent {
   ctxProp: any = 'some value';
 }
 
+@Directive({selector: 'onPrefixedProp'})
+class OnPrefixDir {
+  @Input() onPrefixedProp: any;
```
Comment (mprobst) [inline modules/@angular/core/test/linker/security_integration_spec.ts]: Can you add a test that shows what happens when this works (`on*` imported by current module)? Maybe with an input called `onclick`, so we see what happens when the namespaces collide?

Can we guarantee that a property binding named `on*` never ends up in the DOM, as an attribute?

**Review by mprobst — CHANGES_REQUESTED** [review]
(see above)

File: modules/@angular/core/test/linker/security_integration_spec.ts (hunk 1 above)
Comment (vicb) [inline modules/@angular/core/test/linker/security_integration_spec.ts]: Good catch, what you mentioned is ok but the test was actually wrong.

fixed both.

**Comment by vicb** [comment]
@mpbrost PTAL or @mhevery please validate the changes.

**Comment by vicb** [comment]
got LGTM from @mhevery

File: modules/@angular/core/test/linker/security_integration_spec.ts [hunk 2]
```diff
@@ -43,33 +52,41 @@ function declareTests({useJit}: {useJit: boolean}) {
...
+        // should not throw for inputs starting with "on"
+        let cmp: ComponentFixture<SecuredComponent>;
+        expect(() => cmp = TestBed.createComponent(SecuredComponent)).not.toThrow();
+
+        // must bind to the directive not to the property of the div
+        cmp.detectChanges();
+        const value = cmp.componentInstance.ctxProp;
+        const div = cmp.debugElement.children[0];
+        expect(div.injector.get(OnPrefixDir).onclick).toBe(value);
+        expect(div.nativeElement['onclick']).not.toBe(value);
```
Comment (mprobst) [inline modules/@angular/core/test/linker/security_integration_spec.ts]: This test here would succeed if we accidentally set the property to a different value.

Can you ensure it's not set at all, e.g. not even to `ctxProp` or anything? 
Can you also check div.attributes.getNamedItem?

File: modules/@angular/core/test/linker/security_integration_spec.ts [hunk 3]
```diff
@@ -43,33 +52,41 @@ function declareTests({useJit}: {useJit: boolean}) {
...
+
+        // should not throw for inputs starting with "on"
+        let cmp: ComponentFixture<SecuredComponent>;
+        expect(() => cmp = TestBed.createComponent(SecuredComponent)).not.toThrow();
+
+        // must bind to the directive not to the property of the div
+        cmp.detectChanges();
+        const value = cmp.componentInstance.ctxProp;
+        const div = cmp.debugElement.children[0];
+        expect(div.injector.get(OnPrefixDir).onclick).toBe(value);
```
Comment (mprobst) [inline modules/@angular/core/test/linker/security_integration_spec.ts]: This doesn't seem to have assertions on `onPrefixedProp` – can you remove that, or assert on it?

**Review by mprobst — APPROVED** [review]
LGTM, please fix comments above.

File: modules/@angular/core/test/linker/security_integration_spec.ts (hunk 2 above)
Comment (vicb) [inline modules/@angular/core/test/linker/security_integration_spec.ts]: right, I now set the property.

> Can you also check div.attributes.getNamedItem?

added `expect(div.nativeElement.hasAttribute('onclick')).toEqual(false);`

**Pull request merged**